python/pyarrow/array.pxi

```diff
@@ -281,6 +281,66 @@ cdef _append_array_buffers(const CArrayData* ad, list res):
         _append_array_buffers(ad.child_data[i].get(), res)
 
 
+cdef _reduce_array_data(const CArrayData* ad):
+    """
+    Recursively dissect ArrayData to (pickable) tuples.
+    """
+    cdef size_t i, n
+    assert ad != NULL
+
+    n = ad.buffers.size()
+    buffers = []
+    for i in range(n):
+        buf = ad.buffers[i]
+        buffers.append(pyarrow_wrap_buffer(buf)
+                       if buf.get() != NULL else None)
+
+    children = []
+    n = ad.child_data.size()
+    for i in range(n):
+        children.append(_reduce_array_data(ad.child_data[i].get()))
+
+    return pyarrow_wrap_data_type(ad.type), ad.length, ad.null_count, ad.offset, buffers, children
+
+
+cdef shared_ptr[CArrayData] _reconstruct_array_data(data):
+    """
+    Reconstruct CArrayData objects from the tuple structure generated
+    by _reduce_array_data.
+    """
+    cdef:
+        int64_t length, null_count, offset, i
+        DataType dtype
+        Buffer buf
+        vector[shared_ptr[CBuffer]] c_buffers
+        vector[shared_ptr[CArrayData]] c_children
+
+    dtype, length, null_count, offset, buffers, children = data
+
+    for i in range(len(buffers)):
+        buf = buffers[i]
+        c_buffers.push_back(buf.buffer)
```

Comment:
I think you must check that `buf` is None here, because unfortunately Cython won't tell you that you're doing something wrong: https://github.com/cython/cython/issues/2170

(and yes, it seems in some cases, wrongly dereferencing the attribute gives you the NULL pointer that you are expecting, and lets you think the code is ok :-))

Reply:
(the above is likely to be the reason for the 2.7 crash btw)